### ISSUE TITLE
runtime error

### ISSUE THREAD

**Issue by blueway** [actix/actix-web#437]
when I run:
`examples/async_ex1`

post  url `something_v1`
output:
`thread 'main' panicked at 'overflow when adding duration to time', libcore/option.rs:914:5`

**Comment by fafhrd91** [comment]
works for me, could you post backtrace?

**Comment by blueway** [comment]
```
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::print
             at libstd/sys_common/backtrace.rs:71
             at libstd/sys_common/backtrace.rs:59
   2: std::panicking::default_hook::{{closure}}
             at libstd/panicking.rs:211
   3: std::panicking::default_hook
             at libstd/panicking.rs:227
   4: std::panicking::rust_panic_with_hook
             at libstd/panicking.rs:463
   5: std::panicking::begin_panic_fmt
             at libstd/panicking.rs:350
   6: rust_begin_unwind
             at libstd/panicking.rs:328
   7: core::panicking::panic_fmt
             at libcore/panicking.rs:71
   8: core::option::expect_failed
             at libcore/option.rs:914
   9: <std::time::Instant as core::ops::arith::Add<core::time::Duration>>::add
             at /checkout/src/libcore/option.rs:302
             at libstd/sys/unix/time.rs:46
             at libstd/sys/unix/time.rs:297
             at libstd/time.rs:218
  10: trust_dns_resolver::lookup::Lookup::new_with_max_ttl
             at  cargo/registry/src/github.com-1ecc6299db9ec823/trust-dns-resolver-0.9.0/src/lookup.rs:47
  11: trust_dns_resolver::hosts::read_hosts_conf
             at  cargo/registry/src/github.com-1ecc6299db9ec823/trust-dns-resolver-0.9.0/src/hosts.rs:132
  12: trust_dns_resolver::hosts::Hosts::new
             at  cargo/registry/src/github.com-1ecc6299db9ec823/trust-dns-resolver-0.9.0/src/hosts.rs:35
  13: trust_dns_resolver::resolver_future::ResolverFuture::with_cache::{{closure}}
             at  cargo/registry/src/github.com-1ecc6299db9ec823/trust-dns-resolver-0.9.0/src/resolver_future.rs:154
  14: <futures::future::lazy::Lazy<F, R>>::get
             at cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.23/src/future/lazy.rs:64
  15: <futures::future::lazy::Lazy<F, R> as futures::future::Future>::poll
             at .cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.23/src/future/lazy.rs:82
  16: <alloc::boxed::Box<F> as futures::future::Future>::poll
             at  /.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.23/src/future/mod.rs:113
  17: <actix::fut::FutureWrap<F, A> as actix::fut::ActorFuture>::poll
             at  /.cargo/registry/src/github.com-1ecc6299db9ec823/actix-0.7.3/src/fut/mod.rs:345
  18: <actix::fut::map::Map<A, F> as actix::fut::ActorFuture>::poll
             at  /.cargo/registry/src/github.com-1ecc6299db9ec823/actix-0.7.3/src/fut/map.rs:38
  19: <actix::fut::map_err::MapErr<A, F> as actix::fut::ActorFuture>::poll
             at  /.cargo/registry/src/github.com-1ecc6299db9ec823/actix-0.7.3/src/fut/map_err.rs:38
  20: <alloc::boxed::Box<F> as actix::fut::ActorFuture>::poll
             at  /.cargo/registry/src/github.com-1ecc6299db9ec823/actix-0.7.3/src/fut/mod.rs:275
  21: <actix::contextitems::ActorWaitItem<A>>::poll
             at  /.cargo/registry/src/github.com-1ecc6299db9ec823/actix-0.7.3/src/contextitems.rs:28
  22: <actix::contextimpl::ContextFut<A, C> as futures::future::Future>::poll
             at  /.cargo/registry/src/github.com-1ecc6299db9ec823/actix-0.7.3/src/contextimpl.rs:313
  23: <actix::supervisor::Supervisor<A> as futures::future::Future>::poll
             at  /.cargo/registry/src/github.com-1ecc6299db9ec823/actix-0.7.3/src/supervisor.rs:146
  24: <alloc::boxed::Box<F> as futures::future::Future>::poll
             at  /.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.23/src/future/mod.rs:113
  25: <alloc::boxed::Box<F> as futures::future::Future>::poll
             at  /.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.23/src/future/mod.rs:113
  26: <futures::task_impl::Spawn<T>>::poll_future_notify::{{closure}}
             at  /.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.23/src/task_impl/mod.rs:289
  27: <futures::task_impl::Spawn<T>>::enter::{{closure}}
             at  /.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.23/src/task_impl/mod.rs:363
  28: futures::task_impl::std::set
             at  /.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.23/src/task_impl/std/mod.rs:78
  29: <futures::task_impl::Spawn<T>>::enter
             at  /.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.23/src/task_impl/mod.rs:363
  30: <futures::task_impl::Spawn<T>>::poll_future_notify
             at  /.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.23/src/task_impl/mod.rs:289
  31: <tokio::executor::current_thread::scheduler::Scheduled<'a, U>>::tick
             at  /.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/executor/current_thread/scheduler.rs:352
  32: <tokio::executor::current_thread::scheduler::Scheduler<U>>::tick::{{closure}}
             at  /.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/executor/current_thread/scheduler.rs:332
  33: <tokio::executor::current_thread::Borrow<'a, U>>::enter::{{closure}}::{{closure}}
             at  /.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/executor/current_thread/mod.rs:762
  34: tokio::executor::current_thread::CurrentRunner::set_spawn
             at  /.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/executor/current_thread/mod.rs:794
  35: <tokio::executor::current_thread::Borrow<'a, U>>::enter::{{closure}}
             at  /.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/executor/current_thread/mod.rs:761
  36: <std::thread::local::LocalKey<T>>::try_with
             at /checkout/src/libstd/thread/local.rs:294
  37: <std::thread::local::LocalKey<T>>::with
             at /checkout/src/libstd/thread/local.rs:248
  38: <tokio::executor::current_thread::Borrow<'a, U>>::enter
             at  /.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/executor/current_thread/mod.rs:760
  39: <tokio::executor::current_thread::scheduler::Scheduler<U>>::tick
             at  /.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/executor/current_thread/scheduler.rs:332
  40: <tokio::executor::current_thread::Entered<'a, P>>::tick
             at  /.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/executor/current_thread/mod.rs:611
  41: <tokio::executor::current_thread::Entered<'a, P>>::block_on
             at  /.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/executor/current_thread/mod.rs:495
  42: tokio::runtime::current_thread::runtime::Runtime::block_on::{{closure}}
             at  /.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/runtime/current_thread/runtime.rs:138
  43: tokio::runtime::current_thread::runtime::Runtime::enter::{{closure}}::{{closure}}::{{closure}}::{{closure}}
             at  /.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/runtime/current_thread/runtime.rs:179
  44: tokio_executor::global::with_default::{{closure}}
             at  /.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-executor-0.1.2/src/global.rs:176
  45: <std::thread::local::LocalKey<T>>::try_with
             at /checkout/src/libstd/thread/local.rs:294
  46: <std::thread::local::LocalKey<T>>::with
             at /checkout/src/libstd/thread/local.rs:248
  47: tokio_executor::global::with_default
             at  /.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-executor-0.1.2/src/global.rs:150
  48: tokio::runtime::current_thread::runtime::Runtime::enter::{{closure}}::{{closure}}::{{closure}}
             at  /.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/runtime/current_thread/runtime.rs:177
  49: tokio_timer::timer::handle::with_default::{{closure}}
             at  /.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-timer-0.2.4/src/timer/handle.rs:64
  50: <std::thread::local::LocalKey<T>>::try_with
             at /checkout/src/libstd/thread/local.rs:294
  51: <std::thread::local::LocalKey<T>>::with
             at /checkout/src/libstd/thread/local.rs:248
  52: tokio_timer::timer::handle::with_default
             at  /.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-timer-0.2.4/src/timer/handle.rs:56
  53: tokio::runtime::current_thread::runtime::Runtime::enter::{{closure}}::{{closure}}
             at  /.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/runtime/current_thread/runtime.rs:170
  54: tokio_timer::clock::clock::with_default::{{closure}}
             at  /.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-timer-0.2.4/src/clock/clock.rs:136
  55: <std::thread::local::LocalKey<T>>::try_with
             at /checkout/src/libstd/thread/local.rs:294
  56: <std::thread::local::LocalKey<T>>::with
             at /checkout/src/libstd/thread/local.rs:248
  57: tokio_timer::clock::clock::with_default
             at  /.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-timer-0.2.4/src/clock/clock.rs:119
  58: tokio::runtime::current_thread::runtime::Runtime::enter::{{closure}}
             at  /.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/runtime/current_thread/runtime.rs:169
  59: tokio_reactor::with_default::{{closure}}
             at  /.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-reactor-0.1.2/src/lib.rs:231
  60: <std::thread::local::LocalKey<T>>::try_with
             at /checkout/src/libstd/thread/local.rs:294
  61: <std::thread::local::LocalKey<T>>::with
             at /checkout/src/libstd/thread/local.rs:248
  62: tokio_reactor::with_default
             at  cargo/registry/src/github.com-1ecc6299db9ec823/tokio-reactor-0.1.2/src/lib.rs:214
  63: tokio::runtime::current_thread::runtime::Runtime::enter
             at  cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/runtime/current_thread/runtime.rs:168
  64: tokio::runtime::current_thread::runtime::Runtime::block_on
             at  cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/runtime/current_thread/runtime.rs:136
  65: actix::system::SystemRunner::run
             at  cargo/registry/src/github.com-1ecc6299db9ec823/actix-0.7.3/src/system.rs:200
  66: awc_examples::main
             at async_ex1/src/main.rs:153
  67: std::rt::lang_start::{{closure}}
             at /checkout/src/libstd/rt.rs:74
Panic in Arbiter thread, shutting down system.
```

**Comment by blueway** [comment]
cargo 1.27.0 (1e95190e5 2018-05-27)
rustc 1.27.2 (58cc626de 2018-07-18)

**Comment by fafhrd91** [comment]
@bluejekyll could you help with this, i am not sure where its coming from

```
9: <std::time::Instant as core::ops::arith::Add<core::time::Duration>>::add
             at /checkout/src/libcore/option.rs:302
             at libstd/sys/unix/time.rs:46
             at libstd/sys/unix/time.rs:297
             at libstd/time.rs:218
  10: trust_dns_resolver::lookup::Lookup::new_with_max_ttl
             at  cargo/registry/src/github.com-1ecc6299db9ec823/trust-dns-resolver-0.9.0/src/lookup.rs:47
  11: trust_dns_resolver::hosts::read_hosts_conf
             at  cargo/registry/src/github.com-1ecc6299db9ec823/trust-dns-resolver-0.9.0/src/hosts.rs:132
  12: trust_dns_resolver::hosts::Hosts::new
             at  cargo/registry/src/github.com-1ecc6299db9ec823/trust-dns-resolver-0.9.0/src/hosts.rs:35
  13: trust_dns_resolver::resolver_future::ResolverFuture::with_cache::{{closure}}
             at  cargo/registry/src/github.com-1ecc6299db9ec823/trust-dns-resolver-0.9.0/src/resolver_future.rs:154
  14: <futures::future::lazy::Lazy<F, R>>::ge
```

**Comment by DJMcNab** [comment]
`<std::time::Instant as core::ops::arith::Add<core::time::Duration>>::add` is found here:
https://doc.rust-lang.org/src/std/time.rs.html#214-220.

This function calls `sys::time::Instant::add_duration`, which, as your system appears to be a unix system, is [defined here](https://github.com/rust-lang/rust/blob/8c069ceba81a0fffc1ce95aaf7e8339e11bf2796/src/libstd/sys/unix/time.rs#L166-L171).

This in turn calls `core::u64::checked_add` on the instant's inner time, after the dureation is converted into a time u64 (the inner workings of which involve digging into arcane magic and to identify would probably require me to boot up a virtual machine). The result of this is then `expect`ed

So, as the message says, there was an overflow adding a duration to a time.

Going up the call tree, we see that the function which panics is [this line](https://github.com/bluejekyll/trust-dns/blob/d9963909117e46cce03c3edf716603aa0bbbb31a/resolver/src/lookup.rs#L45), which adds the current system time to a duration converted from seconds ([`pub const MAX_TTL: u32 = 2147483647_u32`](https://github.com/bluejekyll/trust-dns/blob/d9963909117e46cce03c3edf716603aa0bbbb31a/resolver/src/dns_lru.rs#L22)). I can't give any further insight than this, other than to say that I cannot see this failing unless your current system time is greater than `core::u64::MAX/(10^9)`, which would be about on `Sunday, 21 July 2554`. Can you check that calling `Instant::now()` doesn't give you a time greater than that time? @blueway

**Comment by blueway** [comment]
Instant::now() => Instant { tv_sec: 1128, tv_nsec: 267016194 }

**Comment by bluejekyll** [comment]
What architecture is this on? We’re investigating a similar problem on 32bit Linux:

https://github.com/bluejekyll/trust-dns/issues/528

The discussion has been to lower the MAX_TTL to be in 32bit bounds. Is this related to that?

EDIT: I've posted this PR, https://github.com/bluejekyll/trust-dns/pull/542, which will be merged into the 0.9.0 branch, and I'll schedule for a release as 0.9.1.

It would be great if someone could verify this.

**Comment by fafhrd91** [comment]
@blueway closing. Will be fixed by upstream crate
